src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -2198,6 +2245,53 @@ private static bool CanBeMadeAtomic(RegexNode node, RegexNode subsequent, bool i
+                        {
+                            return false;
+                        }
+
+                        bool CharInStartingOrEndingSet(char ch) =>
+                            RegexCharClass.CharInClass(ch, loopStartingSet) || RegexCharClass.CharInClass(ch, loopEndingSet);
+
+                        bool MayOverlapStartingOrEndingSet(string set) =>
+                            RegexCharClass.MayOverlap(set, loopStartingSet) || RegexCharClass.MayOverlap(set, loopEndingSet);
+
+                        switch (subsequent.Kind)
+                        {
+                            case RegexNodeKind.One when !CharInStartingOrEndingSet(subsequent.Ch):
+                            case RegexNodeKind.Set when !MayOverlapStartingOrEndingSet(subsequent.Str!):
+                            case RegexNodeKind.Onelazy or RegexNodeKind.Oneloop or RegexNodeKind.Oneloopatomic when subsequent.M > 0 && !CharInStartingOrEndingSet(subsequent.Ch):
+                            case RegexNodeKind.Setlazy or RegexNodeKind.Setloop or RegexNodeKind.Setloopatomic when subsequent.M > 0 && !MayOverlapStartingOrEndingSet(subsequent.Str!):
+                            case RegexNodeKind.Multi when !CharInStartingOrEndingSet(subsequent.Str![0]):
+                            case RegexNodeKind.End:
+                            case RegexNodeKind.EndZ or RegexNodeKind.Eol when !CharInStartingOrEndingSet('\n'):
+                                return true;
+
+                            case RegexNodeKind.Onelazy or RegexNodeKind.Oneloop or RegexNodeKind.Oneloopatomic when subsequent.M == 0 && !CharInStartingOrEndingSet(subsequent.Ch):
+                            case RegexNodeKind.Setlazy or RegexNodeKind.Setloop or RegexNodeKind.Setloopatomic when subsequent.M == 0 && !MayOverlapStartingOrEndingSet(subsequent.Str!):
+                            case RegexNodeKind.Boundary when node.M > 0 && RegexCharClass.IsKnownWordClassSubset(loopStartingSet) && RegexCharClass.IsKnownWordClassSubset(loopEndingSet):
+                            case RegexNodeKind.NonBoundary when node.M > 0 && (loopStartingSet is RegexCharClass.NotWordClass or RegexCharClass.NotDigitClass) && (loopEndingSet is RegexCharClass.NotWordClass or RegexCharClass.NotDigitClass):
+                            case RegexNodeKind.ECMABoundary when node.M > 0 && (loopStartingSet is RegexCharClass.ECMAWordClass or RegexCharClass.ECMADigitClass) && (loopEndingSet is RegexCharClass.ECMAWordClass or RegexCharClass.ECMADigitClass):
+                            case RegexNodeKind.NonECMABoundary when node.M > 0 && (loopStartingSet is RegexCharClass.NotECMAWordClass or RegexCharClass.NotDigitClass) && (loopEndingSet is RegexCharClass.NotECMAWordClass or RegexCharClass.NotDigitClass):
+                                // The loop can be made atomic based on this subsequent node, but we'll need to evaluate the next one as well.
```

Comment:
Just to see if I'm following this right, this would be for a case like `a*b?abc` where we're deciding if it's okay to turn it into `(?>a*)b?abc`, but we have to check the `a` from `abc` even after we've checked that `b?` doesn't overlap?

Reply:
You got it. It's not safe to make the `a*` atomic in that case, because while it doesn't overlap with `b`, the `b` is nullable, so we also need to check what comes next. What comes next is an `a`, which obviously does overlap, so it can't be made atomic. If it were erroneously changed to be atomic, then on an input like "abc", which should match, it would fail to match, because the `(?>a*)` would end up consuming all the 'a's, and then the subsequent `a` in the pattern would fail to match.